### PR TITLE
Proposal: Conversation API

### PR DESCRIPTION
**Pull request by daixiang0**
It is the initial version to integrate.

The API token here is string for the first version, next we can store it by Crypto component.

### CODE REVIEW

File: 20240618-RCBS-Conversation-building-block.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+## Components
+
+Components in dapr/components-contrib are to be placed in the `conversation` folder and must implement the `Conversation` interface:
+
+```go
+// Conversation offers an interface to perform low-level conversational operations
+type Conversation interface {
+    // Conversate one conversation
+    Conversate(
```
Comment (shubham1172) [inline 20240618-RCBS-Conversation-building-block.md:28]:
```suggestion
    Converse(
```

File: 20240618-RCBS-Conversation-building-block.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        // Context that can be used to cancel the running operation
+        ctx context.Context, 
+        // Name of the model
+        model_name string,
+        // Inputs for the conversation, support multiple input in one time
+        inputs []byte,
+        // Parameters for all custom fields
+        parameters string,
+        // Key of API token
+        key string,
```
Comment (shubham1172) [inline 20240618-RCBS-Conversation-building-block.md:38]: It is a common use case when instances are rate-limited and developers want to load-balance over different instances (or use them in some priority basis). The feature can be kept out of scope for this proposal, but can be worth considering.
```suggestion
        keys []string,
```

**Comment by yaron2** [comment]
Eligible for binding votes: @daixiang0, @yaron2, @artursouza, @mukundansundar, @JoshVanL

File: 20240618-RCBS-Conversation-building-block.md (hunk 2 above)
Comment (daixiang0) [reply 20240618-RCBS-Conversation-building-block.md:38]: Thank you very much, it should be a great feature! Let me take some time to think more about it.

Comment (daixiang0) [reply 20240618-RCBS-Conversation-building-block.md:38]: I add endpoint here, one key for one endpoint.

**Comment by yaron2** [comment]
Proposal looks good. Since a main benefit of this API lies in its abstraction layer, it would be good to have at least two components to begin with for the Alpha stage. Which components would be a good fit for the first stage in your opinion?

**Comment by daixiang0** [comment]
OpenAI is good for start, most LLM support it.

File: 20240618-RCBS-Conversation-building-block.md (hunk 2 above)
Comment (daixiang0) [reply 20240618-RCBS-Conversation-building-block.md:40]: Update endpoint to endpoints, support rr load balancing for now.

**Comment by yaron2** [comment]
It would also be good to investigate how we can integrate with LangChain and offer higher level integrations for it with existing APIs like state, pub/sub and workflows in addition to providing native Dapr features like authentication, authorization and resiliency policies.

**Comment by daixiang0** [comment]
This [article](https://www.octomind.dev/blog/why-we-no-longer-use-langchain-for-building-our-ai-agents) may help us think better to integrate Conversation API with existing APIs.

The first step we can easily move is integrate store API to use LLM API keys.

File: 20240618-RCBS-Conversation-building-block.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  // An array of results.
+  repeated ConversationResult outputs = 1;
+}
+```
+
+## HTTP APIs
+
+The HTTP APIs are same with the gRPC APIs：
+
+`POST /v1.0/conversation/[component]/converse` -> Conversate
```
Comment (yaron2) [inline 20240618-RCBS-Conversation-building-block.md:105]: Can you please add the JSON request/responses?

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: 20240618-RCBS-Conversation-building-block.md (hunk 3 above)
Comment (daixiang0) [reply 20240618-RCBS-Conversation-building-block.md:78]: Sure.

**Comment by yaron2** [comment]
+1 binding

File: 20240618-RCBS-Conversation-building-block.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  // Endpoints for the model service, co-work with load balanace policy.
+  repeated string endpoints = 4;
+  // Name of the model.
+  string model = 5;
+  // Key of API token
+  string key = 6;
+  // Load balancing policy for endpoints.
+  LoadBalancingPolicy policy = 7;
```
Comment (artursouza) [inline 20240618-RCBS-Conversation-building-block.md:59]: Can the policy be moved to the component metadata instead?

File: 20240618-RCBS-Conversation-building-block.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    // Round robin policy.
+    ROUNDROBIN = 0;
+  }
+
+  // Endpoints for the model service, co-work with load balanace policy.
+  repeated string endpoints = 4;
+  // Name of the model.
+  string model = 5;
+  // Key of API token
+  string key = 6;
```
Comment (artursouza) [inline 20240618-RCBS-Conversation-building-block.md:57]: Can this be moved to the component metadata?

File: 20240618-RCBS-Conversation-building-block.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+  enum LoadBalancingPolicy {
+    // Round robin policy.
+    ROUNDROBIN = 0;
+  }
+
+  // Endpoints for the model service, co-work with load balanace policy.
+  repeated string endpoints = 4;
+  // Name of the model.
+  string model = 5;
```
Comment (artursouza) [inline 20240618-RCBS-Conversation-building-block.md:55]: Can this be moved to the component metadata?

File: 20240618-RCBS-Conversation-building-block.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+  // The metadata passing to conversation components
+
+  enum LoadBalancingPolicy {
+    // Round robin policy.
+    ROUNDROBIN = 0;
+  }
+
+  // Endpoints for the model service, co-work with load balanace policy.
+  repeated string endpoints = 4;
```
Comment (artursouza) [inline 20240618-RCBS-Conversation-building-block.md:53]: Can this be moved to the component metadata?

**Review by artursouza — CHANGES_REQUESTED** [review]

File: 20240618-RCBS-Conversation-building-block.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+// (Existing Dapr service)
+service Dapr {
+  // Conversate.
+  rpc Converse(stream ConversationRequest) returns (stream ConversationResponse);
+}
+
+// ConversationRequest is the request object for Conversation.
+message ConversationRequest {
+  // Inputs for the conversation, support multiple input in one time.
+  repeated string inputs = 1;
```
Comment (JoshVanL) [inline 20240618-RCBS-Conversation-building-block.md:38]: What is an input? A token, word, sentence, line, paragraph?
Why would I want to have multiple inputs?

I would be helpful to have some example usage (and responses) using this API.

File: 20240618-RCBS-Conversation-building-block.md (hunk 4 above)
Comment (daixiang0) [reply 20240618-RCBS-Conversation-building-block.md:59]: Sure.

File: 20240618-RCBS-Conversation-building-block.md (hunk 8 above)
Comment (daixiang0) [reply 20240618-RCBS-Conversation-building-block.md:38]: Please check the end of doc, each input means a question, here we support ask multiple questions in one time.

**Comment by daixiang0** [comment]
@artursouza has moved those into metadata and update the HTTP example.

**Comment by artursouza** [comment]
I changed the proposal to move the metadata to component. I also added `conversationContext` to allow tracking of the chat and recover a session.

**Comment by daixiang0** [comment]
@artursouza thank you very much!

**Comment by daixiang0** [comment]
+1 binding

**Review by yaron2 — APPROVED** [review]

**Pull request merged**